libs/language-server/src/lib/validation/checks/constraint-definition.ts

```diff
@@ -45,4 +57,42 @@ export function checkConstraintExpression(
   }
 
   checkExpressionSimplification(expression, props);
+
+  const foundFittingFreeVariable =
+    iterateSubExpressionBreadthFirst(expression, (exp): true | undefined => {
+      if (isValueKeywordLiteral(exp)) {
+        if (valueTypePropertyName !== undefined) {
+          props.validationContext.accept(
+            'error',
+            "Inline constraint definitions must not contain the value keyword literal. Use the valuetype's property instead",
+            {
+              node: exp,
+            },
+          );
+          return undefined;
+        }
+        return true;
+      } else if (isReferenceLiteral(exp)) {
+        if (valueTypePropertyName === undefined) {
+          return undefined;
+        }
+        return valueTypePropertyName === exp.value.ref?.name ? true : undefined;
+      } else if (isFreeVariableLiteral(exp)) {
+        assertUnreachable(exp);
+      }
+      return undefined;
+    }) ?? false;
```

Comment:
It might make sense to extract. Returning a boolean here threw me off a little :)